### PR TITLE
Wip/6.0 re-enable tests from org.hibernate.test.join package

### PR DESCRIPTION
**Pull request by dreab8**

### CODE REVIEW

File: hibernate-core/src/test/java/org/hibernate/orm/test/join/Person.java [hunk 1]
```diff
@@ -30,7 +30,6 @@
 @Entity
 @Inheritance(strategy = InheritanceType.SINGLE_TABLE)
 @DiscriminatorColumn( name = "person_type" )
-@DiscriminatorValue( value = "P" )
```
Comment (beikov) [inline hibernate-core/src/test/java/org/hibernate/orm/test/join/Person.java:33]: I guess this change was by accident?

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 2]
```diff
@@ -3311,6 +3311,25 @@ protected final void renderCommaSeparated(Iterable<? extends SqlAstNode> express
...
+				expression.accept( this );
+				separator = COMA_SEPARATOR;
+			}
+		}
+	}
+
+	private boolean isNullValue(SqlAstNode expression) {
+		if(expression instanceof QueryLiteral){
+			QueryLiteral literal = (QueryLiteral) expression;
+			return literal.getLiteralValue().equals( "null" ) || literal.getLiteralValue().equals( "not null" );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:3328]: Wouldn't this also be triggered if I did a query like `from Entity a where a.name in ('null', 'not null')` which is completely unrelated? IMO we should introduce special `QueryLiteral` objects or a special expression class for which we can do an identity check here.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 3]
```diff
@@ -4000,13 +4020,46 @@ else if ( !supportsRowValueConstructorSyntaxInInList() ) {
...
+			if ( nullExpressions > 0 && !areOnlyNullExpressions ) {
+				appendSql( OPEN_PARENTHESIS );
+			}
 			inListPredicate.getTestExpression().accept( this );
-			if ( inListPredicate.isNegated() ) {
-				appendSql( " not" );
+
+			if ( nullExpressions > 0 ) {
+				if ( !areOnlyNullExpressions ) {
+				}
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:4039]: Unnecessary if?

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 4]
```diff
@@ -4000,13 +4020,46 @@ else if ( !supportsRowValueConstructorSyntaxInInList() ) {
...
+			if ( !areOnlyNullExpressions ) {
+				if ( negated ) {
+					appendSql( " not" );
+				}
+				appendSql( " in (" );
+				renderNonNullCommaSeparated( listExpressions );
+				appendSql( CLOSE_PARENTHESIS );
+			}
+			if ( nullExpressions > 0 && !areOnlyNullExpressions ) {
+				appendSql( CLOSE_PARENTHESIS );
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:4061]: Move this into the if above and eliminate the second conjunction.

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 5]
```diff
@@ -3311,6 +3311,25 @@ protected final void renderCommaSeparated(Iterable<? extends SqlAstNode> express
 		}
 	}
 
+	protected final void renderNonNullCommaSeparated(Iterable<? extends SqlAstNode> expressions) {
+		String separator = NO_SEPARATOR;
+		for ( SqlAstNode expression : expressions ) {
+			if( !isNullValue( expression ) ){
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:3317]: Formatting

File: hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java [hunk 6]
```diff
@@ -4000,13 +4020,46 @@ else if ( !supportsRowValueConstructorSyntaxInInList() ) {
 			}
 		}
 		else {
+			int nullExpressions = 0;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/sql/ast/spi/AbstractSqlAstTranslator.java:4023]: How about we do all of this only when the LHS has a model part type of `java.lang.Class`?

**Review by beikov — CHANGES_REQUESTED** [review]

File: hibernate-core/src/test/java/org/hibernate/orm/test/join/Person.java (hunk 1 above)
Comment (dreab8) [reply hibernate-core/src/test/java/org/hibernate/orm/test/join/Person.java:33]: yes it was :)

File: hibernate-core/src/main/java/org/hibernate/persister/entity/SingleTableEntityPersister.java [hunk 7]
```diff
@@ -983,10 +988,32 @@ private Predicate createDiscriminatorPredicate(
...
+				final NullnessPredicate nullnessPredicate = new NullnessPredicate( sqlExpression );
+				if ( hasNonNull ) {
+					nullnessPredicate.isNegated();
+				}
+				junction.add( nullnessPredicate );
+				junction.add( p );
+				return junction;
 			}
-			return new InListPredicate( sqlExpression, values );
+			return p;
```
Comment (beikov) [inline hibernate-core/src/main/java/org/hibernate/persister/entity/SingleTableEntityPersister.java:1016]:
```suggestion
			final Predicate p = new InListPredicate( sqlExpression, values );
			if ( hasNull || hasNonNull ) {
				final Junction junction = new Junction(
						Junction.Nature.DISJUNCTION
				);
				// This essentially means we need to select everything, so we don't need a predicate at all
				if ( hasNull && hasNonNull ) {
					return junction;
				}
				final NullnessPredicate nullnessPredicate = new NullnessPredicate( sqlExpression, hasNonNull );
				junction.add( nullnessPredicate );
				junction.add( p );
				return junction;
			}
			return p;
```

Comment (dreab8) [reply hibernate-core/src/main/java/org/hibernate/persister/entity/SingleTableEntityPersister.java]: Thanks @beikov changed it

**Review by beikov — APPROVED** [review]

**Pull request merged**